tools/pyi/gen_pyi.py

```diff
@@ -176,54 +176,44 @@ def sig_for_ops(opname: str) -> List[str]:
     else:
         raise Exception("unknown op", opname)
 
-def generate_named_tuples(funcs: Sequence[PythonSignatureGroup]) -> Dict[str, str]:
-    namedtuples: Dict[str, str] = {}
-    for sig_group in funcs:
-        named_tuple = sig_group.signature.returns.named_tuple_pyi()
-        if named_tuple is not None:
-            tuple_name, tuple_def = named_tuple
-            if tuple_name in namedtuples:
-                assert namedtuples[tuple_name] == tuple_def
-            else:
-                namedtuples[tuple_name] = tuple_def
-    return namedtuples
-
-def generate_type_hints(funcs: Sequence[PythonSignatureGroup], is_tensor: bool = False) -> List[str]:
-    """generate_type_hints(funcs, is_tensor=False)
-
-    Generates type hints for the declarations pertaining to the function
-    :attr:`funcs` are the func from the parsed native_functions.yaml.
-    The :attr:`is_tensor` flag indicates whether we are parsing
-    members of the Tensor class (true) or functions in the
-    `torch` namespace (default, false).
-    """
+def generate_named_tuples(sig_group: PythonSignatureGroup, namedtuples: Dict[str, str]) -> None:
```

Comment:
nit: if you can, try to avoid nonlocal mutation like what's going on in this function. The moment you start have to start reasoning across function calls to figure out what is mutating, it is just a bad look. You could for example, maintain the old structure and have it return a `Tuple[str, str]`, and then assign into `namedtuples` at the call site, keeping the mutation local.

Reply:
The original codegen had an assertion against global state that was worth keeping in (for every named_tuple we encounter, check that its definition doesn't collide with an existing one).

Totally agreed with what you're saying- I'm fixing that by getting rid of `generate_named_tuples` entirely and moving the logic to the call-site.